### DISCUSSION TITLE
development environment

### DISCUSSION THREAD

**Discussion by math-GH** [FreshRSS/FreshRSS#3882] [General]
Hi,

which development environment do you use? Could you suggest something?

I use Visual Studio Code and I am looking f.e. for the best scss->css compiler (it is not so easy to find a good one that complies with tab as indent and not with 2 spaces)

**Comment by Frenzie** [comment]
The style guide doesn't apply where it doesn't make sense (e.g., CSS files generated by SASS, auto-generated stuff in general, third party libs). 

https://github.com/FreshRSS/FreshRSS/blob/a473b90ff9533687e1347aed0d3b9ceaa5e66aab/.stylelintignore#L1-L4

**Comment by math-GH** [comment]
Another important thing to use here in the project:

PHP Codesniffer:
https://github.com/squizlabs/PHP_CodeSniffer

(see install instructions on Github: first download it and then install it with Composer (do not forget to install Composer ;) )).

The "phpcs" plugin for Visual Studio Code helps while developing
![grafik](https://user-images.githubusercontent.com/1645099/137628769-b9b504f3-8664-4580-a40f-f344e28de98a.png)

The plugin shows you within the Editor the wrong styled code.

![grafik](https://user-images.githubusercontent.com/1645099/137628935-270a4a1e-4cf2-45b0-ac80-c90637552341.png)


As @Alkarex mentioned: Before commiting/create an PR run this 2 lines in your terminal:
```
phpcbf -p -s .
phpcs -p -s .
```

**Comment by Frenzie** [comment]
NB Running such tools locally is advisable before you submit a PR but
normally the CI will do it for you.[1] (So alternatively you can just push
to your branch on GH - just don't submit it as a PR yet.)

[1] Normally meaning the past couple years but broken right this moment.

**Comment by Alkarex** [comment]
Hello,
Here are some of the relevant extensions I use in Visual Studio Code:

* https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker
* https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig
* https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint (I need to move the default settings from JSHint to ESLint)
* https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint
* https://marketplace.visualstudio.com/items?itemName=bmewburn.vscode-intelephense-client
* https://marketplace.visualstudio.com/items?itemName=ValeryanM.vscode-phpsab
* https://marketplace.visualstudio.com/items?itemName=foxundermoon.shell-format
* https://marketplace.visualstudio.com/items?itemName=timonwong.shellcheck

**Comment by math-GH** [comment]
For work at the themes I use https://marketplace.visualstudio.com/items?itemName=yutent.scss-to-css to convert SCSS files to CSS.  Just save the SCSS file and it will generate directly the CSS file (configuration: use 1 `tab` instead of `space` )

(I have to investigate to create RTL CSS files from the SCSS file)

**Reply by Frenzie** [reply]
You can use `make rtl`.

https://github.com/FreshRSS/FreshRSS/blob/9e9a6c3838f13d04c5ca126c4379978fe8f2fa95/Makefile#L160-L162

> (I have to investigate to create RTL CSS files from the SCSS file)

If you did it directly from SCSS you'd have to ignore those directories for rtlcss. That sounds to me like more trouble than it's worth compared to having one tool for everything.

**Reply by math-GH** [reply]
I have zero knowledge to use `make` on a Windows machine.
I installed `make` as it is described here: https://www.youtube.com/watch?v=taCJhnBXG_w

But cannot run it. Parsing error on `&&`:
![grafik](https://user-images.githubusercontent.com/1645099/138956399-408440fe-4783-43ce-aa57-a0b5ede05c04.png)

**Reply by Frenzie** [reply]
Oh, that's too bad. The Makefile is just intended to make things simpler because it comes with autocomplete that a shell script does not. We're not really using any of Make's features that make it, well, Make. As you can see all that particular command does is to run:

```
rtlcss -d p/themes/ && find p/themes/ -type f -name '*.rtl.rtl.css' -delete
```

`rtlcss -d p/themes/` generates the RTL CSS. The second part deletes the extra files it generates on top of that. (Cf. `https://github.com/MohammadYounes/rtlcss/issues/64`.) In PowerShell, the equivalent would be something like `gci -r -fi *.rtl.rtl.css | remove-item`

**Reply by Alkarex** [reply]
@mathContao On Windows, [take a look at WSL](https://docs.microsoft.com/windows/wsl/install). It gives you a Linux terminal (e.g. Ubuntu) with a nice integration, and everything will Just work™

**Reply by Frenzie** [reply]
That's true. In the past I've used Cygwin a bit but WSL is quite a bit more useful.
